lib/ace/ext/papeeria/latex_behaviour.coffee

```diff
@@ -0,0 +1,226 @@
+define((require, exports, module) ->
+  Behaviour = require("ace/mode/behaviour").Behaviour
+  COMMENT_TYPE = "comment"
+  ESCAPE_TYPE = "escape"
+  KEYWORD_TYPE = "keyword"
+  STORAGE_TYPE = "storage"
+  MATH_TYPE = "math"
+  EQUATION_TYPE = "equation"
+  RPAREN_TYPE = "rparen"
+  CORRESPONDING_CLOSING = {
+      '(': ')'
+      '[': ']'
+      '{': '}'
+  }
+
+
+  # If we're on the start of a line, but the current token is of a type "comment", we're not actually
+  # inside a comment, just on a start of it. That's the reason for `column != 0` check
+  isCommentToken = (token, column) -> column != 0 and token? and token.type.indexOf(COMMENT_TYPE) > -1
+
+
+  isEscapedInsertion = (token, column) -> (
+      token? and
+      (
+          token.type.indexOf(ESCAPE_TYPE) > -1 or
+          token.type.indexOf(KEYWORD_TYPE) > -1 or
+          token.type.indexOf(STORAGE_TYPE) > -1
+      ) and
+      column - token.start == 1
+  )
+
+
+  isInEquation = (session, row, column) ->
```

Comment:
We use `LatexParsingContext` to find out our _context_, and auto-closing behaviors are not based entirely on _context_.